File: drake/util/resolveCenterOfPressure.m
% Computes center of pressure and normal torque at computed center of
% pressure given force, torque and a contact plane defined by a normal and
% a point  on the plane. All input variables should be defined in the same
% frame, and the output will be expressed in this frame as well.
% Multiple CoPs can be computed at once by passing in multi-column inputs.
%
% @param torque torque exerted at a given contact plane
% @param force force exerted at a given contact plane
% @param normal unit normal to the contact plane. Direction (inward or
% outward) defines sign of normal torque according to the right hand rule
% @param point_on_contact_plane an arbitrary point on the contact plane.
%
% @retval cop the center of pressure. If there is no force normal to the
% surface, nan(3,n) will be returned for n-column input.
% @retval normal_torque_at_cop scalar normal component of the torque
% at the computed cop. If there is no force normal to the surface, nan is
% returned (since there is no cop at which to compute the normal torque).

function [cop, normal_torque_at_cop] = resolveCenterOfPressure(torque, force, normal, point_on_contact_plane)
if any(abs(dot(normal, normal) - 1) > 1e-12)
  error('cols of normal should be a unit vectors')
end

n = size(torque, 2);
cop = nan(3, n);
normal_torque_at_cop = nan(1, n);

fz = dot(normal, force);
cop_exists = abs(fz) > 1e-12;
torque = torque(:, cop_exists);
force = force(:, cop_exists);
normal = normal(:, cop_exists);
point_on_contact_plane = point_on_contact_plane(:, cop_exists);
fz = fz(:, cop_exists);

torque_at_point_on_contact_plane = torque - cross(point_on_contact_plane, force);
normal_torque_at_point_on_contact_plane = dot(normal, torque_at_point_on_contact_plane);
tangential_torque = torque_at_point_on_contact_plane - bsxfun(@times, normal, normal_torque_at_point_on_contact_plane);
cop_compact = bsxfun(@rdivide, cross(normal, tangential_torque), fz) + point_on_contact_plane;
cop(:, cop_exists) = cop_compact;

if nargout > 1
  torque_at_cop = torque - cross(cop_compact, force);
  normal_torque_at_cop(:, cop_exists) = dot(normal, torque_at_cop);
end

end

File: drake/multibody/resolve_center_of_pressure.cc
#include "drake/multibody/resolve_center_of_pressure.h"

#include <limits>
#include <stdexcept>

template <typename DerivedTorque, typename DerivedForce, typename DerivedNormal,
          typename DerivedPoint>
std::pair<Eigen::Vector3d, double> resolveCenterOfPressure(
    const Eigen::MatrixBase<DerivedTorque>& torque,
    const Eigen::MatrixBase<DerivedForce>& force,
    const Eigen::MatrixBase<DerivedNormal>& normal,
    const Eigen::MatrixBase<DerivedPoint>& point_on_contact_plane) {
  // TODO(tkoolen): implement multi-column version
  using Eigen::Vector3d;

  if (std::abs(normal.squaredNorm() - 1.0) > 1e-12) {
    throw std::runtime_error(
        "Drake:resolveCenterOfPressure:BadInputs: normal should be a unit "
        "vector");
  }

  Vector3d cop;
  double normal_torque_at_cop;

  double fz = normal.dot(force);
  bool cop_exists = std::abs(fz) > 1e-12;

  if (cop_exists) {
    auto torque_at_point_on_contact_plane =
        torque - point_on_contact_plane.cross(force);
    double normal_torque_at_point_on_contact_plane =
        normal.dot(torque_at_point_on_contact_plane);
    auto tangential_torque = torque_at_point_on_contact_plane -
                             normal * normal_torque_at_point_on_contact_plane;
    cop = normal.cross(tangential_torque) / fz + point_on_contact_plane;
    auto torque_at_cop = torque - cop.cross(force);
    normal_torque_at_cop = normal.dot(torque_at_cop);
  } else {
    cop.setConstant(std::numeric_limits<double>::quiet_NaN());
    normal_torque_at_cop = std::numeric_limits<double>::quiet_NaN();
  }
  return std::pair<Vector3d, double>(cop, normal_torque_at_cop);
}

template std::pair<Eigen::Matrix<double, 3, 1, 0, 3, 1>, double>
resolveCenterOfPressure<Eigen::Map<Eigen::Matrix<double, 3, 1, 0, 3, 1> const,
                                   0, Eigen::Stride<0, 0> >,
                        Eigen::Map<Eigen::Matrix<double, 3, 1, 0, 3, 1> const,
                                   0, Eigen::Stride<0, 0> >,
                        Eigen::Map<Eigen::Matrix<double, 3, 1, 0, 3, 1> const,
                                   0, Eigen::Stride<0, 0> >,
                        Eigen::Map<Eigen::Matrix<double, 3, 1, 0, 3, 1> const,
                                   0, Eigen::Stride<0, 0> > >(
    Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<double, 3, 1, 0, 3, 1> const, 0,
                                 Eigen::Stride<0, 0> > > const&,
    Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<double, 3, 1, 0, 3, 1> const, 0,
                                 Eigen::Stride<0, 0> > > const&,
    Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<double, 3, 1, 0, 3, 1> const, 0,
                                 Eigen::Stride<0, 0> > > const&,
    Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<double, 3, 1, 0, 3, 1> const, 0,
                                 Eigen::Stride<0, 0> > > const&);
template
    std::pair<Eigen::Matrix<double, 3, 1, 0, 3, 1>, double>
    resolveCenterOfPressure<Eigen::Matrix<double, 3, 1, 0, 3, 1>,
                            Eigen::Matrix<double, 3, 1, 0, 3, 1>,
                            Eigen::Matrix<double, 3, 1, 0, 3, 1>,
                            Eigen::Matrix<double, 3, 1, 0, 3, 1> >(
        Eigen::MatrixBase<Eigen::Matrix<double, 3, 1, 0, 3, 1> > const&,
        Eigen::MatrixBase<Eigen::Matrix<double, 3, 1, 0, 3, 1> > const&,
        Eigen::MatrixBase<Eigen::Matrix<double, 3, 1, 0, 3, 1> > const&,
        Eigen::MatrixBase<Eigen::Matrix<double, 3, 1, 0, 3, 1> > const&);
template
    std::pair<Eigen::Matrix<double, 3, 1, 0, 3, 1>, double>
    resolveCenterOfPressure<
        Eigen::Block<Eigen::Matrix<double, 6, 1, 0, 6, 1>, 3, 1, false>,
        Eigen::Block<Eigen::Matrix<double, 6, 1, 0, 6, 1>, 3, 1, false>,
        Eigen::Matrix<double, 3, 1, 0, 3, 1>,
        Eigen::Matrix<double, 3, 1, 0, 3, 1> >(
        Eigen::MatrixBase<Eigen::Block<Eigen::Matrix<double, 6, 1, 0, 6, 1>, 3,
                                       1, false> > const&,
        Eigen::MatrixBase<Eigen::Block<Eigen::Matrix<double, 6, 1, 0, 6, 1>, 3,
                                       1, false> > const&,
        Eigen::MatrixBase<Eigen::Matrix<double, 3, 1, 0, 3, 1> > const&,
        Eigen::MatrixBase<Eigen::Matrix<double, 3, 1, 0, 3, 1> > const&);
